### ISSUE TITLE
Mkdocs build is very slow

### ISSUE THREAD

**Issue by mfisher87** [nsidc/earthaccess#545]
It's largely the jupyter notebooks. Mkdocs currently lacks caching features we would need to fix this: https://github.com/danielfrg/mkdocs-jupyter/issues/161

**Comment by betolink** [comment]
This would be nice indeed! maybe we can contribute upstream? Just a note that this is mostly for when we are working on a single notebook, for the whole documentation I think we can pass `--dirtyreload` to mkdocs to avoid re-executing the whole site. https://github.com/mkdocs/mkdocs/issues/1074

**Comment by betolink** [comment]
Correct, I don't think readthedocs has a partial build workflow, the hope is that if context is persisted between builds maybe `jupyter-cache` can speed things up.

**Comment by mfisher87** [comment]
I don't know how others feel about priority, but I feel the sooner we get this done, the better for our contributor base. If you're interested in working on this, I think that'd be great! :)

**Comment by mfisher87** [comment]
Sounds good! I don't know how to proceed myself; perhaps the best place to start is to ask what work needs to be done in the mkdocs-jupyter repository (issue linked at the top of this issue) to get this task going. There's a chance an expert won't be able to advise on this work. I'd start here, where mkdocs-jupyter plugin is configuring nbconvert to execute the notebook based on the user's config (represented by the `execute` boolean variable):

https://github.com/danielfrg/mkdocs-jupyter/blob/ded7803bac2166426ccc238c87c83bf14ffe0701/src/mkdocs_jupyter/plugin.py#L123

Luis' idea is to use jupyter-cache with nbconvert so that only changed cells will re-execute. This is where I get a bit foggy (see [jupyter-cache Python API](https://jupyter-cache.readthedocs.io/en/latest/using/api.html)). How feasible is this? What are the downsides (e.g. adding cache data to the repo)? Is it exactly what we want? I personally think it would be OK if we cached at the notebook file level instead of at the cell level. I.e. if I changed 1 notebook, 1 notebook re-renders, and if I changed 0, 0 will re-render. Most docs changes don't touch the notebooks.

When we're talking about notebook caching, it's important to recognize that notebooks already "cache" their outputs internally. But that data that's stored internally is dependent on how the notebook is executed. If a human re-runs a cell multiple times or runs cells out-of-order, the numbers won't be sequential. We don't want folks reading our docs to see that kind of confusing stuff. This is why we execute the notebooks from scratch every time we build the docs, but I think we could do better.

What if we only execute notebooks that aren't already executed with sequential output numbers? Then we'd need a process to ensure notebooks are committed in that state to avoid excessive re-execution. We can also potentially take inspiration from Quarto's "freeze" feature which caches execution outputs. I'm not sure how it works with notebooks, though.

If you want to pursue this, and I think it sounds like a really fun and rewarding project, come join us on the new Jupyter Zulip chat! Welcoming experts who can guide this work and answer questions are there. Specifically, I'd start in the "ask-anything" channel and go from there. https://jupyter.zulipchat.com

**Comment by chuckwondo** [comment]
As @betolink alluded to above, we can pass a flag to `mkdocs`, but it's named `--dirty`, not `--dirtyreload`, so when locally running the mkdocs server, we can immensely shorten our iteration time by running one of the following commands.

(**Assumption:** you have followed one of the setup approaches detailed in `docs/contributing/development.md`.)

If you are using `nox`:

```plain
nox -s serve-docs -- --dirty 
```

If you not using `nox`, but instead are using either a regular Python venv or a conda env, and your environment is activated, you can run the following instead:

```plain
mkdocs serve --dirty
```

Either way, it will still take several minutes for the mkdocs server to be ready because it will re-execute/render all of the notebooks every time the server is _started_.

However, once the server is ready, it will hot reload _only_ files that you change (or add) while the server is running. This _immensely_ improves local rendering, as it will happen in only a few seconds for `*.md` files, and for notebooks, it will re-execute _only_ modified notebooks (rather than _every_ notebook), which may still take on the order of 10s of seconds, depending upon the notebook, but is still far better than having to wait for _everything_ to re-execute.

**Comment by chuckwondo** [comment]
Given my previous comment, I suggest that the solution to this issue is to simply update our "contributing" docs to include the details I gave above.

**Comment by mfisher87** [comment]
Playing with the docs today to help merge #899 , and this is really getting in the way :(

**Comment by asteiker** [comment]
Related workaround: #1095

**Comment by saberbrasher** [comment]
Bringing this back to life and putting it in the backlog as I don't _think_ it has been addressed yet... Mentioned in #1095 as Amy linked, as well as #1194. If I am understanding correctly, it boils down to lengthy tutorials... but I'm still not sure there has been consensus on if that sort of content should stay in our docs (does it make sense for a library?) or not.

**Comment by Sherwin-14** [comment]
I am pretty sure a migration from MKDocs as planned, as the project seems to be imploding on itself.

**Comment by mfisher87** [comment]
I don't know how realistic it is that we'll actually migrate any time soon :( It would be a good bit of work!

I don't think of this as a lengthy tutorial. I think of it more of a side-note (admonition/callout) in a "build the docs locally" how-to
